subsys/net/openthread/rpc/server/ot_rpc_commissioning_server.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <ot_rpc_ids.h>
+#include <ot_rpc_types.h>
+#include <ot_rpc_common.h>
+
+#include <nrf_rpc/nrf_rpc_serialize.h>
+#include <nrf_rpc/nrf_rpc_cbkproxy.h>
+
+#include <zephyr/net/openthread.h>
+
+#include <openthread/link.h>
+
+static void ot_thread_discover_cb(otActiveScanResult *result, void *context, uint32_t callback_slot)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	struct ot_rpc_active_scan_result pkt_data;
+	size_t cbor_buffer_size = result ? sizeof(pkt_data) + 1 : 1;
+
+	cbor_buffer_size += 10; /* for context and callback slot */
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, cbor_buffer_size);
+
+	if (result) {
+		memcpy(pkt_data.ext_addr, result->mExtAddress.m8, OT_EXT_ADDRESS_SIZE);
+		memcpy(pkt_data.nwk_name, result->mNetworkName.m8, OT_NETWORK_NAME_MAX_SIZE + 1);
+		memcpy(pkt_data.ext_pan_id, result->mExtendedPanId.m8, OT_EXT_PAN_ID_SIZE);
+		memcpy(pkt_data.steering_data, result->mSteeringData.m8,
+		       OT_STEERING_DATA_MAX_LENGTH);
+		pkt_data.pan_id = result->mPanId;
+		pkt_data.joiner_udp_port = result->mJoinerUdpPort;
+		pkt_data.channel = result->mChannel;
+		pkt_data.rssi = result->mRssi;
+		pkt_data.lqi = result->mLqi;
+		pkt_data.version = result->mVersion;
+		pkt_data.is_native = result->mIsNative;
+		pkt_data.discover = result->mDiscover;
+		pkt_data.is_joinable = result->mIsJoinable;
+
+		nrf_rpc_encode_buffer(&ctx, &pkt_data, sizeof(pkt_data));
+	} else {
+		nrf_rpc_encode_null(&ctx);
+	}
+
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)context);
+	nrf_rpc_encode_uint(&ctx, callback_slot);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_THREAD_DISCOVER_CB, &ctx, ot_rpc_decode_void,
+				NULL);
+}
+
+NRF_RPC_CBKPROXY_HANDLER(ot_thread_discover_cb_encoder, ot_thread_discover_cb,
```

Comment:
Before I finish the review - if we start using the callback proxy for OpenThread group, we won't be able to easily write POSIX unit tests for the server side (which I started to work on). I think we have two choices:
1. Port the callback proxy assembly to x64 and x86. Kinda meh but possible :)
2. Do not use the callback proxy in OT commands taking advantage of the fact that all OT APIs that take callback parameters take an extra context as well. The server can allocate a pair `<client's callback, client's context>` and pass that pair as a context to the actual OT API. It's more portable and more flexible than the callback proxy approach.

What are your thoughts @canisLupus1313 @alxelax?

Reply:
> 
> 1. Port the callback proxy assembly to x64 and x86. Kinda meh but possible :)

Kind if disagree with this statement. Unit tests check serialization\deserialization of API. Callback is out of scope of this. All callback proxy API can be mocked, so unit tests will be possible to run on any architecture.

Reply:
Yes, but we still have to be able to build this file and test that the callback results in a proper serialization.

Reply:
It is possible to exclude it for unit tests. I do not see unresolvable issues. If OpenThread paradigm of context usage is changed, then ot rpc will require reimplementation.
Even more Bluetooth RPC should be tested as well. Potentially with unit tests too. There is no context in Bluetooth API.

Reply:
I'm also not saying this is unresolvable. I just mentioned the limitations this implementation imposes but like with other things we can discuss live and choose the best option.